AST: teach Decl::getLoc() to return serialized source loc.

nkcsgexi opened the pull request:
When Decl::getLoc() is called upon a serialized AST and the
serialized source location is available, we lazily open the
external buffer and return a valid SourceLoc instance pointing
into the buffer.

nkcsgexi committed:
AST: teach Decl::getLoc() to return serialized source loc.

When Decl::getLoc() is called upon a serialized AST and the
serialized source location is available, we lazily open the
external buffer and return a valid SourceLoc instance pointing
into the buffer.

nkcsgexi commented:
@swift-ci please smoke test

slavapestov commented on lib/AST/Decl.cpp:
```diff
@@ -487,7 +487,26 @@ SourceLoc Decl::getLoc() const {
+    if (auto Locs = File->getBasicLocsForDecl(this)) {
+      auto &SM = getASTContext().SourceMgr;
+      auto Loc = Locs->Loc;
+      return SM.getLocFromExternalSource(Locs->SourceFilePath, Loc.Line,
```
Should this be cached in the decl?

nkcsgexi replied:
That's not a bad idea. Will update!

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi committed:
AST: cache serialized source locations in Decl

jrose-apple commented on include/swift/AST/Decl.h:
```diff
@@ -666,6 +666,13 @@ class alignas(1 << DeclAlignInBits) Decl {
+    SourceLoc StartLoc;
+    SourceLoc EndLoc;
+  };
+  mutable Optional<CachedExternalSourceLocs> SerializedLocsCache;
```
:-( It's unfortunate that this bloats the size of every Decl. Maybe we should just always have these three source locations, and bits to say whether they're computed, and then we can cache the normal lookups too? Or have them behind an indirection allocated on the ASTContext, so that it's only one word instead of three, and only if there are actual valid locations.

jrose-apple commented on lib/AST/Decl.cpp:
```diff
@@ -482,12 +486,49 @@ case DeclKind::ID: return cast<ID##Decl>(this)->getLocFromSource();
+  auto Locs = File->getBasicLocsForDecl(this);
+  if (!Locs.hasValue()) {
+    return Result;
+  }
```
One trick you can do here for the common case of "no locs" is to have them all share a single `static const Decl::CachedExternalSourceLocs NoLocs;`. That avoids an allocation.

nkcsgexi replied:
Good point! Fixed in the newest commit.

jrose-apple replied:
Ah, I suggest *not* doing it in Decl.h; there's no reason to expose it outside of this function. The `const` is nice too because then it can go in the const data section. You don't actually need the *struct* to be mutable if it's stored out-of-line.

nkcsgexi committed:
SourceManager: avoid using additional buffer Id map for external source buffers

SourceManager may outlive ASTContext where external source file paths are managed.
Therefore, we should use the identifier from the memory buffer as the key instead of
using a pointer from an ASTContext.

nkcsgexi committed:
AST: use a static allocated NullLocs to indicate no serialized source locations for a Decl, NFC

nkcsgexi committed:
lit: move -ignore-module-source-info flag to swift_test_options

jrose-apple commented on lib/AST/Decl.cpp:
```diff
@@ -410,6 +410,10 @@ template<typename Class>
 
+template<typename Class>
+inline char checkSourceLocType(SourceLoc (Class::*)(bool) const);
+inline TwoChars checkSourceLocType(SourceLoc (Decl::*)(bool) const);
```
It looks like you didn't end up using these? (They're the same as what's above.)

nkcsgexi replied:
They are difference for taking `bool` as function argument. We need these for `getLoc(bool)` and the above ones for `getLocFromSource()`.

jrose-apple replied:
Oh, whoops. My mistake!

jrose-apple commented on test/lit.cfg:
```diff
@@ -752,7 +752,7 @@ if run_vendor == 'apple':
+        "-target %s %s %s -ignore-module-source-info" %
         (config.variant_triple, resource_dir_opt, mcp_opt))
     target_options_for_mock_sdk = (
         "-target %s %s %s" %
```
Oops, why isn't this based on `target_options`. 😭 It belongs here too, *and* in the non-Apple sections!

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi committed:
AST: unblock linux build by providing explicit constructor

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi commented:
@swift-ci please smoke test

jrose-apple commented on include/swift/AST/Decl.h:
```diff
@@ -667,6 +667,15 @@ class alignas(1 << DeclAlignInBits) Decl {
+    SourceLoc StartLoc;
+    SourceLoc EndLoc;
+    // We need to keep this user-defined constructor build on linux.
+    CachedExternalSourceLocs() {}
```
Wow, that definitely shouldn't be necessary. (I believe you though.) Do you still remember what the error message was?

nkcsgexi replied:
It seems the explicit constructor isn't necessary. All we need is to add `{}` after the `CachedExternalSourceLocs` data, see here: https://stackoverflow.com/questions/7411515/why-does-c-require-a-user-provided-default-constructor-to-default-construct-a

jrose-apple replied:
Ah, I guess that makes sense. Zero-initializing wouldn't be correct, so you actually have to call the constructor explicitly. Or something.

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi committed:
sourcekitd-test: ignore .swiftsourcefile when invoking sourcekitd from sourcekitd-test

nkcsgexi committed:
AST: avoid looking up serialized locations when decl is context-free

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi commented:
@swift-ci please test
